internal/providers/aws/aws.go

```diff
@@ -73,3 +75,49 @@ func Init(f *resource.Fetcher) error {
 	f.S3RegionHint = regionHint
 	return nil
 }
+
+// FetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its configured, else
+// it will fall back to `IMDSv1`. By default, the value of url is an IMDSv1 endpoint
+// i.e. http://169.254.169.254/2019-10-01/user-data.
+func FetchFromAWSMetadata(u url.URL, opts resource.FetchOptions, f *resource.Fetcher) ([]byte, error) {
+	userdataLatestURL := url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "latest",
+	}
+	token, err := fetchAWSIMDSV2Token(userdataLatestURL, f)
+	if err == resource.ErrIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is unavailable; falling back to IMDSv1")
+	} else if err != nil {
+		return nil, err
+	} else if len(token) > 0 && err == nil {
+		opts.Headers = http.Header{
+			"X-aws-ec2-metadata-token": []string{token},
+		}
+		u.Path = path.Join(userdataLatestURL.Path, "user-data")
+	}
+	return f.FetchToBuffer(u, opts)
+}
+
+// fetchAWSIMDSV2Token fetches a session token from an EC2 instance (if the
+// instace is configured to use `IMDSv2`), otherwise, it will return an error.
+func fetchAWSIMDSV2Token(u url.URL, f *resource.Fetcher) (string, error) {
+	u.Path = path.Join(u.Path, "api/token")
+	opts := resource.FetchOptions{
+		Headers: http.Header{
+			"x-aws-ec2-metadata-token-ttl-seconds": []string{"21600"},
+		},
+		HTTPVerb: "PUT",
+	}
+	token, err := f.FetchToBuffer(u, opts)
+	if err != nil {
+		if err == resource.ErrNotFound {
```

Comment:
nit: I'd be inclined to structure this as `if err == resource.ErrNotFound {} else if err != nil {}`.